File: xcrypto/ed25519.go
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

//go:build darwin

package xcrypto

import (
	"errors"
	"strconv"

	"github.com/microsoft/go-crypto-darwin/internal/cryptokit"
)

const (
	// publicKeySizeEd25519 is the size, in bytes, of public keys as used in crypto/ed25519.
	publicKeySizeEd25519 = 32
	// privateKeySizeEd25519 is the size, in bytes, of private keys as used in crypto/ed25519.
	privateKeySizeEd25519 = 64
	// signatureSizeEd25519 is the size, in bytes, of signatures generated and verified by crypto/ed25519.
	signatureSizeEd25519 = 64
	// seedSizeEd25519 is the size, in bytes, of private key seeds. These are the private key representations used by RFC 8032.
	seedSizeEd25519 = 32
)

// PublicKeyEd25519 represents an Ed25519 public key.
type PublicKeyEd25519 []byte

// PrivateKeyEd25519 represents an Ed25519 private key.
type PrivateKeyEd25519 []byte

func (k PrivateKeyEd25519) Public() PublicKeyEd25519 {
	publicKey := make([]byte, publicKeySizeEd25519)
	copy(publicKey, k[seedSizeEd25519:])
	return PublicKeyEd25519(publicKey)
}

// GenerateKeyEd25519 generates a new Ed25519 private key.
func GenerateKeyEd25519() PrivateKeyEd25519 {
	pkeyPriv := make([]byte, privateKeySizeEd25519)
	cryptokit.GenerateKeyEd25519(pkeyPriv)
	return pkeyPriv
}

func NewPrivateKeyEd25519(priv []byte) (PrivateKeyEd25519, error) {
	if len(priv) != privateKeySizeEd25519 {
		panic("ed25519: bad private key length: " + strconv.Itoa(len(priv)))
	}
	return NewPrivateKeyEd25519FromSeed(priv[:seedSizeEd25519])
}

func (k PrivateKeyEd25519) Bytes() ([]byte, error) {
	return k, nil
}

func NewPublicKeyEd25519(pub []byte) (PublicKeyEd25519, error) {
	if len(pub) != publicKeySizeEd25519 {
		panic("ed25519: bad public key length: " + strconv.Itoa(len(pub)))
	}
	pkey := make([]byte, publicKeySizeEd25519)
	result := cryptokit.NewPublicKeyEd25519(pkey, pub)
	if result != 0 {
		return nil, errors.New("failed to create Ed25519 public key")
	}
	return pkey, nil
}

func (k PublicKeyEd25519) Bytes() ([]byte, error) {
	return k, nil
}

// NewPrivateKeyEd25519FromSeed calculates a private key from a seed. It will panic if
// len(seed) is not [SeedSize]. RFC 8032's private keys correspond to seeds in this
// package.
// NewPrivateKeyEd25519FromSeed creates an Ed25519 private key from a seed.
func NewPrivateKeyEd25519FromSeed(seed []byte) (PrivateKeyEd25519, error) {
	if len(seed) != seedSizeEd25519 {
		panic("ed25519: bad seed length: " + strconv.Itoa(len(seed)))
	}
	pkey := make([]byte, privateKeySizeEd25519)
	result := cryptokit.NewPrivateKeyEd25519FromSeed(pkey, seed)
	if result != 0 {
		return nil, errors.New("failed to generate Ed25519 key from seed")
	}
	return pkey, nil
}

// SignEd25519 signs the message with priv and returns a signature.
func SignEd25519(priv PrivateKeyEd25519, message []byte) ([]byte, error) {
	sig := make([]byte, signatureSizeEd25519)
	result := cryptokit.SignEd25519(priv, message, sig)
	if result < 0 {
		switch result {
		case -1:
			return nil, errors.New("invalid inputs to SignEd25519")
		case -2:
			return nil, errors.New("failed to reconstruct private key")
		case -3:
			return nil, errors.New("failed to sign the message")
		case -4:
			return nil, errors.New("signature buffer too small")
		default:
			return nil, errors.New("unknown error in SignEd25519")
		}
	}
	return sig, nil
}

// VerifyEd25519 reports whether sig is a valid signature of message by pub.
func VerifyEd25519(pub PublicKeyEd25519, message, sig []byte) error {
	result := cryptokit.VerifyEd25519(pub, message, sig)
	switch result {
	case 1:
		return nil // Valid signature
	case 0:
		return errors.New("ed25519: invalid signature")
	case -1:
		return errors.New("invalid inputs to VerifyEd25519")
	case -2:
		return errors.New("failed to reconstruct public key")
	default:
		return errors.New("unknown error in VerifyEd25519")
	}
}

File: cryptokit/Sources/CryptoKitSrc/ed25519.swift
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import CryptoKit
import Foundation

var publicKeySizeEd25519 = 32
var signatureSizeEd25519 = 64
var seedSizeEd25519 = 32

@_cdecl("generateKeyEd25519")
public func generateKeyEd25519(keyPointer: UnsafeMutablePointer<UInt8>) -> Void {
    // Generate a private key using CryptoKit
    let privateKey = Curve25519.Signing.PrivateKey()

    // Extract the raw representation of the private key
    let privateKeyData = privateKey.rawRepresentation

    // Allocate memory for the private key data and copy it
    privateKeyData.copyBytes(to: keyPointer, count: privateKeyData.count)
    privateKey.publicKey.rawRepresentation.copyBytes(to: keyPointer + publicKeySizeEd25519, count: publicKeySizeEd25519)
}

@_cdecl("newPrivateKeyEd25519FromSeed")
public func newPrivateKeyEd25519FromSeed(
    keyPointer: UnsafeMutablePointer<UInt8>,
    seedPointer: UnsafePointer<UInt8>
) -> Int {
    // Copy the seed into a Data object
    let seedData = Data(bytes: seedPointer, count: seedSizeEd25519)

    // Generate the private key from the seed
    let privateKey: Curve25519.Signing.PrivateKey
    do {
        privateKey = try Curve25519.Signing.PrivateKey(rawRepresentation: seedData)
    } catch {
        // Key generation failed
        return 1
    }

    // Extract the raw representation of the private key
    let privateKeyData = privateKey.rawRepresentation

    // Allocate memory for the private key data and copy it
    privateKeyData.copyBytes(to: keyPointer, count: privateKeyData.count)
    privateKey.publicKey.rawRepresentation.copyBytes(to: keyPointer + publicKeySizeEd25519, count: publicKeySizeEd25519)
    return 0
}

@_cdecl("newPublicKeyEd25519")
public func newPublicKeyEd25519(
    keyPointer: UnsafeMutablePointer<UInt8>,
    pubPointer: UnsafePointer<UInt8>
) -> Int {
    // Copy the public key bytes into a Data object
    let pubData = Data(bytes: pubPointer, count: seedSizeEd25519)

    do {
        // Create the public key from the raw representation
        let publicKey = try Curve25519.Signing.PublicKey(rawRepresentation: pubData)

        // Store the public key in memory and return the pointer
        publicKey.rawRepresentation.copyBytes(to: keyPointer, count: publicKeySizeEd25519)
        return 0
    } catch {
        // Failed to create the public key
        return 1
    }
}

@_cdecl("signEd25519")
public func signEd25519(
    privateKeyPointer: UnsafePointer<UInt8>,
    messagePointer: UnsafePointer<UInt8>?,
    messageLength: Int,
    sigBuffer: UnsafeMutablePointer<UInt8>?
) -> Int {
    guard let sigBuffer = sigBuffer else {
        return -1  // Invalid inputs
    }

    // Convert the raw private key back to the seed (32 bytes)
    let privateKeySeed = Data(bytes: privateKeyPointer, count: seedSizeEd25519)

    // Reconstruct the private key
    let privateKey: Curve25519.Signing.PrivateKey
    do {
        privateKey = try Curve25519.Signing.PrivateKey(rawRepresentation: privateKeySeed)
    } catch {
        return -2  // Failed to reconstruct private key
    }

    // Convert the message to Data
    let messageData: Data
    if let messagePointer = messagePointer, messageLength > 0 {
        messageData = Data(bytes: messagePointer, count: messageLength)
    } else {
        messageData = Data()  // Empty message
    }

    // Sign the message
    let signature: Data
    do {
        signature = try privateKey.signature(for: messageData)
    } catch {
        return -3  // Failed to sign the message
    }

    // Ensure the buffer is large enough
    guard signature.count == signatureSizeEd25519 else {
        return -4  // Buffer too small
    }

    // Copy the signature to the buffer
    signature.copyBytes(to: sigBuffer, count: signature.count)

    return signature.count  // Return the number of bytes written
}

@_cdecl("verifyEd25519")
public func verifyEd25519(
    publicKeyPointer: UnsafePointer<UInt8>,
    messagePointer: UnsafePointer<UInt8>?,
    messageLength: Int,
    sigPointer: UnsafePointer<UInt8>?
) -> Int {
    guard let sigPointer = sigPointer else {
        return -1  // Invalid inputs
    }

    // Convert the raw public key back to a Data object
    let publicKeyData = Data(bytes: publicKeyPointer, count: publicKeySizeEd25519)

    // Reconstruct the public key
    guard let publicKey = try? Curve25519.Signing.PublicKey(rawRepresentation: publicKeyData) else {
        return -2  // Error: failed to reconstruct public key
    }

    // Convert the message and signature to Data
    let rawMessage: Data
    if let messagePointer = messagePointer, messageLength > 0 {
        rawMessage = Data(bytes: messagePointer, count: messageLength)
    } else {
        rawMessage = Data()  // Empty message
    }
    let signatureData = Data(bytes: sigPointer, count: signatureSizeEd25519)

    // Verify the signature
    let isValid = publicKey.isValidSignature(signatureData, for: rawMessage)
    return isValid ? 1 : 0  // Return 1 for valid, 0 for invalid
}
